design/secure-communication-design.md

```diff
@@ -0,0 +1,79 @@
+# Secure Communication between Contour and Envoy
+
+_Status_: Draft, in review.
+
+This document outlines how to secure the communication between Contour
+and Envoy containers.
+
+## Goals
+
+- Achieve secure communication between Contour and Envoy
+- Authenticate Envoy connecting to Contour
+- Rotating and invalidating the credentials
+
+## Non-goals
+
+- Support for dynamic credentials
+- Support for use of Secret Discovery Service for credential management
+
+## Background
+
+Currently, Contour and Envoy containers are co-located within a pod.
+Current security model is relying on the fact that Contour exposes the management
+server on loopback interface which prevents any unauthorized access from outside
+the pod.
+
+With the need to move away from the [current deployment pattern of co-locating
+Contour and Envoy][0], it is necessary to secure the communication channel between
+the containers.
+
+## High level design
+
+- Use mutual TLS to secure communication channel between Contour and Envoy
+- Bootstrap phase of Contour will generate necessary TLS certificates
+
+## Detailed design
+
+### Bootstrapping process
+
+By default, mutual TLS is enabled to secure the communication between Contour and Envoy
+containers. If `contour bootstrap` is passed with the parameter `--disable-tls`,
+the communication between Contour and Envoy will not be secured.
+
+The bootstrap phase will create three pairs of the TLS certifications. The details about
+each pair of certificates are mentioned below:
+1. The first pair is used by Contour to serve HTTPS (named `contour.crt`, `contour.key`)
+2. The second pair is used by Envoy to authenticate with Contour (named `envoy.crt`, `envoy.key`)
+3. The last pair serves as the CA certificate which signs the above certificates (named `ca.crt`, `ca.key`)
+
+These certificates will be placed under `/config/tls/`. The bootstrap configuration yaml
+of Envoy will have [`tls_context`][1] added as part of `cluster` object.
+
+### Setting up secure communication channel
+
+1. Contour container will pick the certificates `contour.crt`, `contour.key` and `ca.crt` from
```

Comment:
Is this going to be hard coded in the config passed to the envoy container?

Reply:
I have updated the design to make use of Kubernetes secrets instead of using the TLS certificate as files.